scalalib/src/mill/scalalib/RunModule.scala

```diff
@@ -46,27 +48,37 @@ trait RunModule extends WithZincWorker {
   def mainClass: T[Option[String]] = None
 
   /**
-   * Same as [[allLocalMainClasses]], but only for modules with a custom
-   * JVM version configured
-   */
-  def allLocalMainClasses0: T[Seq[String]] = Task { Seq.empty[String] }
-
-  /**
-   * All main classes detected in this module that can serve as program entrypoints
+   * All main classes detected in this module that can serve as program entry-points.
    */
   def allLocalMainClasses: T[Seq[String]] = Task {
-    if (zincWorker().javaHome().isEmpty) allLocalMainClasses0()
-    else {
-      Jvm.callProcess(
-        mainClass = "mill.scalalib.worker.DiscoverMainClassesMain",
-        classPath = zincWorker().classpath().map(_.path).toVector,
-        mainArgs = Seq(localRunClasspath().map(_.path).mkString(",")),
-        javaHome = zincWorker().javaHome().map(_.path),
-        stdin = os.Inherit,
-        stdout = os.Pipe,
-        cwd = Task.dest
-      ).out.lines()
+    classgraphWorker().discoverMainClasses(localRunClasspath().map(_.path))
+  }
+
+  def classgraphWorkerClasspath: T[Seq[PathRef]] = T {
+    zincWorker().defaultResolver().classpath(Seq(
+      Dep.millProjectModule("mill-scalalib-classgraph-worker")
+    ))
+  }
+
+  def classgraphWorker: Worker[ClassgraphWorker] = Task.Worker {
+    val cp = classgraphWorkerClasspath().map(_.path)
+    Task.log.debug(s"Create classloader for classpath: ${cp}")
+    val cl = mill.util.Jvm.createClassLoader(cp, getClass().getClassLoader())
+    val klass = "mill.scalalib.classgraph.impl.ClassgraphWorkerImpl"
+    Task.log.debug(s"Loading class: ${klass}")
+    val cls = cl.loadClass(klass)
+    Task.log.debug(s"Instantiating class: ${klass}")
+    val worker = cls.getConstructor().newInstance().asInstanceOf[ClassgraphWorker]
+
+    object Worker extends ClassgraphWorker with AutoCloseable {
+      override def discoverMainClasses(classpath: Seq[Path])(implicit ctx: Ctx): Seq[String] =
+        worker.discoverMainClasses(classpath)
+
+      override def close(): Unit = {
+        cl.close()
+      }
     }
+    Worker
```

Comment:
This code is pretty awkward.

* I'd just make it one big `new ClassgraphWorker extends AutoCloseable{}` and put everything inside.
* The logging can probably be removed; anyone who wants to debug stuff can add it, but it really makes the code much harder to read
* The naming needs to be improved. `klass` vs `cls` is super confusing

Reply:
That means I need two "worker" classes. The `AutoClosable` handling obviously needs to happen outside the real `ClassgraphWorker`, but since the `ClassgraphWorker` is instantiated from the managed classloader, it can't (or shouldn't) provide the cleanup logic.

I can cleanup further...

Reply:
fixed